### PR TITLE
[22.05] emacs: use withPgtk option more

### PR DESCRIPTION
**Pull request by leungbk**
###### Description of changes

Backport of https://github.com/NixOS/nixpkgs/pull/189929

Cherry-picked from commit 6d0e82f11aeaef74f8f9adcf18c1c77df8c8e2c6

There was a conflict; in this cherry-picked version of the commit, `withGTK3` now has its default value of `true` unchanged, whereas in the commit on `master` branch, `withGTK3` had its previous value of `false` changed to `withPgtk`. I think this is a reasonable resolution to the conflict, but let me know if you believe otherwise.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adisbladis** [comment]
> There was a conflict; in this cherry-picked version of the commit, withGTK3 now has its default value of true unchanged, whereas in the commit on master branch, withGTK3 had its previous value of false changed to withPgtk. I think this is a reasonable resolution to the conflict, but let me know if you believe otherwise.

LGTM!

I don't understand why this is causing Darwin rebuilds? Any ideas?

**Comment by leungbk** [comment]
> I don't understand why this is causing Darwin rebuilds? Any ideas?

Since `withGTK3` defaults to true on this branch, I'm guessing it's this line:

https://github.com/leungbk/nixpkgs/blob/43354079bcf9f9aef863cc1c23d70eaca11db3fa/pkgs/applications/editors/emacs/generic.nix#L139

File: pkgs/applications/editors/emacs/generic.nix [hunk 1]
```diff
@@ -134,7 +135,9 @@ let emacs = stdenv.mkDerivation (lib.optionalAttrs nativeComp {
     ++ lib.optionals withImageMagick [ imagemagick ]
     ++ lib.optionals (stdenv.isLinux && withX) [ m17n_lib libotf ]
     ++ lib.optional (withX && withGTK2) gtk2-x11
-    ++ lib.optionals (withX && withGTK3) [ gtk3-x11 gsettings-desktop-schemas ]
+    ++ lib.optional (withX && withGTK3) gtk3-x11
+    ++ lib.optional withGTK3 gsettings-desktop-schemas
```
Comment (leungbk) [inline pkgs/applications/editors/emacs/generic.nix:139]: I suppose I should change this in a follow-up cherry to something like

```suggestion
    ++ lib.optional (!withNS && withGTK3) gsettings-desktop-schemas
```

I don't have a good understanding of macOS so I could be wrong. Does this seem reasonable? If so, I'll make an additional PR onto master branch and cherry-pick that onto this branch.

**Comment by leungbk** [comment]
> > I don't understand why this is causing Darwin rebuilds? Any ideas?
> 
> Since `withGTK3` defaults to true on this branch, I'm guessing it's this line:
> 
> https://github.com/leungbk/nixpkgs/blob/43354079bcf9f9aef863cc1c23d70eaca11db3fa/pkgs/applications/editors/emacs/generic.nix#L139
> 
> edit: no wait, that wouldn't explain why Darwin rebuilds were triggered on the PR targeting `master` though...

I thought about this more and looked at the x86 Darwin log (https://logs.nix.ci/?key=nixos/nixpkgs.189929&attempt_id=5d961d1e-c6b7-4c56-98b7-458ae6a090d0) for the other PR targeting master, where some of the default arguments to the `emacs/generic.nix` expression are different. The corresponding GitHub label says that 1-10 rebuilds occurred for Darwin, but I don't see anything in the log that can be reasonably described as a rebuild; the Emacs package itself seems to be unchanged, while the test doesn't run because of a failed assertion. Does the non-running test from that other PR count as a rebuild?

The different default I mentioned before would explain why there are *more* rebuilds here on the stable branch, though if there is a non-trivial rebuild for Darwin that occurs in both branches, then I still can't explain it.

**Pull request merged**